rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,163 @@
+
+However, it is possible for "Two-Way" mode to emulate the other three in a way that is also mostly-backwards-compatible with the current manual mode. Furthermore, a singular IaSQL database would be able to have different "parts" of the database emulating different modes at the same time.
+
+To accomplish this, let's assume under normal operations, there is a singular apply/sync-combined command to perform the "Two-Way"-style action. It works similarly to the current sync and apply, but also uses the audit log to decide for each record which behavior to follow. This piece can be operated in a "manual" style, invoked explicitly, but we'll instead invoke it through a cron-like interval, every few minutes, to make the automatic mode. In this way, there is no special syntax for users to work with normally. They just `SELECT/INSERT/UPDATE/DELETE` records they see and eventually this is reflected in their cloud account. This should also work with migration systems without issues, though in an eventually-consistent fashion.
+
+To emulate manual mode would basically be to temporarily turn off that cron, let the user make whatever changes they want to batch together, and then re-enable that cron. Because we can't actually support the current `iasql_apply` for the three `code-` services, and because entering this batching mode that could potentially be unwound, it looks a *lot* like a transaction, so the proposed new syntax would be to emulate that:
+
+```sql
+CALL iasql_begin(); -- Disables the cron. Wrap a function call inside of a procedure so we can avoid the confusing `SELECT iasql_begin();` syntax here.
+
+-- Do some DB changes
+
+SELECT * FROM iasql_preview(); -- Because there's no distinction between apply and sync, this function name is simpler than before
+
+-- If we don't like what we saw
+
+CALL iasql_rollback(); -- Executes `iasql_sync`, then re-enables the cron
+
+-- If we like what we saw
+
+CALL iasql_commit(); -- Re-enables the cron, but also performs a blocking execution of the new logic so a `SELECT` on the newly-created records has the desired cloud IDs
+```
+
+Backing this would be a minor addition to the existing audit log. If we have the new cron logic and the rollback/commit calls add rows to the audit log when it starts and ends, they can check if the most recent of these records is a `start` one, implying that the engine is currently working on applying and synchronizing changes with the cloud, and would alter their behavior slightly. The `cron` would simply early exit in this case, as it implies there is an existing cron that has gone overtime, and so it should wait until the next execution of the cron to try again. `iasql_begin` would block until a paired `end` record exists and then continue with its work. `iasql_rollback` and `iasql_commit` should therefore never see this state after `iasql_begin` is called, but for defensiveness they should as well block until an `end` record exists and then do their own work.
```

Comment:
I think that approach simplifies in some ways and makes more complicated in others.

It seems to eliminate the need for a locking mechanism in the "normal" usage of Two-Way mode, but we would still need some sort of locking mechanism if the user calls `iasql_begin()`, so that basically shifts the lock up from the database and into the engine. That *can* work at the moment when we have one process doing the work, but we would have to re-centralize the lock once we go multi-process.

Meanwhile, the cron job could actually be powered by `pg_cron` and run on the database itself, and we could have it simply check the state of the audit log to decide if it should run or not and not involve the engine at all in this process, and `iasql_begin()`'s implementation in that scenario could be pure SQL as well, simply inserting a `start_commit` into the audit log once the most recent `start`/`end` pair has completed (then the same logic used to deal with long-running cron jobs would also block the cron job when the user wants the activity to be transactional).

Finally, we can take a look at how these two deal with various failure modes:


```
+------------------+-----------------------------------+-----------------------------------+
| Failure Mode     | Cron                              | Event                             |
+------------------+-----------------------------------+-----------------------------------+
| Stuck job >45min | The graphile worker library kills | The graphile worker library kills |
|                  | the job after 45 minutes and then | the job after 45 minutes and then |
|                  | the cron starts work again, which | no event for the next run is      |
|                  | may resolve correctly this time   | triggered, which causes this DB   |
|                  | or it may be stuck again, but any | to be soft-locked until there is  |
|                  | new changes by the user to the DB | an engine redeploy.               |
|                  | will be considered.               |                                   |
+------------------+-----------------------------------+-----------------------------------+
| Engine dies      | The cron keeps trying and failing | Nothing happens at all.           |
|                  | to call the engine. The cron SQL  |                                   |
|                  | function inserts start and end    |                                   |
|                  | audit records immediately after   |                                   |
|                  | each other, and does nothing.     |                                   |
+------------------+-----------------------------------+-----------------------------------+
| Database dies    | Nothing happens at all.           | The inner cron of the engine will |
|                  |                                   | keep trying to communicate to the |
|                  |                                   | DB and fail. This may or may not  |
|                  |                                   | be handled correctly by our code. |
+------------------+-----------------------------------+-----------------------------------+
| Engine deployed  | The cron creates the graphile job | The two engines both have their   |
|                  | and one of the two parallel       | own inner cron unaware of each    |
|                  | versions of the engine takes the  | other and very likely they will   |
|                  | work (as this approach requires a | both try to execute at the same   |
|                  | SQL RPC call to start the engine) | time for 7-10 minutes. What they  |
|                  | and the other doesn't work on it. | will do is undefined unless we    |
|                  | If it is the old version that     | thoroughly test it, but likely    |
|                  | takes it, it may fail and no work | they will "fight" with each other |
|                  | will happen for 45 minutes, then  | undoing each other's work and     |
|                  | it will continue.                 | potentially corrupting the DB.    |
+------------------+-----------------------------------+-----------------------------------+
```

Basically, I think the event approach has the harder-to-handle failure scenarios, though there are definitely problems that come up with both.

Reply:
I turned this explanation into a section of the RFC.